### PR TITLE
Persistence v6

### PR DESCRIPTION
**Pull request by edersondisouza**
v6:
No more file size checking before use;
Previous topic enforced `sol_util_file` use and improvements/fixing:
                  `sol_util_load_file_fd_string()` to read a file descriptor - useful for efivars reading, that need to discard unused parts of file;
                   Fixed off-by-one reading;
Using sol_str_table to lookup which functions to use;
Error checking on `fwrite()`.

v5:
Instead of error packets on no previous stored value, now we use a 'default_value';
RESET port to reset node to its default_value;
As the code was reorganised, the commit about 'alloca' as droped as its content merged into previous commit.

v4:
No changes. Just to bring semaphore back.

v3:
Return error if packet requested size differs from persisted size;
Fixed make allyesconfig failure;
Added option 'error_on_no_value' - when true (default false) it'll send an error packet informing that no previous value was found upon node opening.

v2:
No more 'fs_dir_path'. If storage is 'fs', 'name' can contain a fully fledged path to a file;
Using more sol_buffer;
Using additional read and write to deal with efivars 'attributes' - avoid allocations and simplifying code;
Depending on Linux for now, thus avoiding some ifdefs;

Adding 'efivars' support to Soletta. I decided to reuse 'fs' node type, but renamed it to 'persistence' - however name suggestions are welcome.
From user's perspective, it's like a dictionary: give a name to a packet and save it. User can choose storage type, currently 'efivars' and 'fs'. In case it chooses 'fs', [s]he must choose a 'fs_dir_path' - dir where persisted files will reside - but I'm open to suggestion on how to improve this: I didn't like having to make user define it for every persistence node, but maybe is more a question of config file.

### CODE REVIEW

File: src/modules/flow/persistence/persistence.c [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+
+static int
+persist_reset(struct persist_data *mdata, struct sol_flow_node *node)
+{
+    void *value;
+
+    if (mdata->packet_data_size)
+        value = alloca(mdata->packet_data_size);
+
+    value = mdata->node_get_default_fn(node);
```
Comment (lpereira) [inline src/modules/flow/persistence/persistence.c]: Value is allocated in the previous if but overwritten here. Why?

File: src/modules/flow/persistence/persistence.c [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+    char *fs_dir_path;
+    char *storage;
+
+    struct sol_flow_packet *(*packet_new_fn)(const struct persist_data *data);
+    int (*packet_data_get_fn)(const struct sol_flow_packet *packet, void *value_ptr);
+    int (*packet_send_fn)(struct sol_flow_node *node);
+    void *(*node_get_default_fn)(struct sol_flow_node *node);
+
+    int (*storage_write)(const char *name, struct sol_buffer *buffer);
+    int (*storage_read)(const char *name, struct sol_buffer *buffer);
```
Comment (lpereira) [inline src/modules/flow/persistence/persistence.c]: Put these two function pointers in a structure, have a pointer to the structure here, and make the str_table point to fs/efivars struct.

File: src/modules/flow/persistence/persistence.c [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+
+    return persist_open(node, data, opts->storage, opts->name);
+}
+
+static void
+persist_string_close(struct sol_flow_node *node, void *data)
+{
+    struct persist_string_data *mdata = data;
+
+    free(mdata->default_value);
```
Comment (lpereira) [inline src/modules/flow/persistence/persistence.c]: Is default_value properly initialized when not passed down from the opts struct?

Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: Is it possible that opts struct does not have default_value?

File: src/modules/flow/persistence/persistence.c (hunk 1 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: Drugs...

File: src/modules/flow/persistence/persistence.c (hunk 2 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: ok

File: src/modules/flow/persistence/persistence.c (hunk 3 above)
Comment (edersondisouza) [inline src/modules/flow/persistence/persistence.c]: I've tested it and found no issue.

**Comment by edersondisouza** [comment]
v7 on #668

**Pull request closed without merging**